dart/lib/src/gurmukhi.dart

```diff
@@ -0,0 +1,195 @@
+extension GurmukhiExtensions on String {
+  /// Remove vishraams from Gurmukhi text.
+  ///
+  /// ```dart
+  /// 'ਸਭਨਾ ਜੀਆ ਕਾ, ਇਕੁ ਦਾਤਾ; ਸੋ. ਮੈ ਵਿਸਰਿ ਨ ਜਾਈ ॥੫॥'.removeVishraams();
+  /// // 'ਸਭਨਾ ਜੀਆ ਕਾ ਇਕੁ ਦਾਤਾ ਸੋ ਮੈ ਵਿਸਰਿ ਨ ਜਾਈ ॥੫॥'
+  /// ```
+  String removeVishraams() => replaceAll(RegExp('[.,;]'), '');
+
+  /// Get first letters from Gurmukhi text.
+  ///
+  /// ```dart
+  /// 'ਜਿਸਨੋ ਕ੍ਰਿਪਾ ਕਰਹਿ ਤਿਨਿ ਨਾਮ ਰਤਨੁ ਪਾਇਆ ॥'.firstGurmukhiLetters();
+  /// // ['ਜ', 'ਕ', 'ਕ', 'ਤ', 'ਨ', 'ਰ', 'ਪ']
+  /// ```
+  Iterable<String> firstGurmukhiLetters() {
+    return split(' ')
+        .where((word) => word.isNotEmpty && _isGurmukhiChar(word.codeUnitAt(0)))
+        .map((word) => word[0]);
+  }
+
+  /// Split Gurmukhi text into letters.
+  ///
+  /// ```dart
+  /// 'ਕੈਸੀ ਆਰਤੀ ਹੋਇ ॥'.splitGurmukhi();
+  /// // ['ਕੈ', 'ਸੀ', ' ', 'ਆ', 'ਰ', 'ਤੀ', ' ', 'ਹੋ', 'ਇ', ' ', '॥']
+  /// ```
+  ///
+  /// Enable [splitVirama] to split chars joined by virama.
+  ///
+  /// ```dart
+  /// 'ਅੰਮ੍ਰਿਤ'.splitGurmukhi();
+  /// // ['ਅੰ', 'ਮ੍ਰਿ', 'ਤ']
+  /// 'ਅੰਮ੍ਰਿਤ'.splitGurmukhi(splitVirama: true);
+  /// // ['ਅੰ', 'ਮ੍', 'ਰਿ', 'ਤ']
+  /// ```
+  ///
+  /// Enable [extensions] to support SantLipi modifiers.
+  ///
+  /// ```dart
+  /// 'ਸ꠴ਯਾਮ'.splitGurmukhi(extensions: true);
+  /// // ['ਸ', '꠴ਯਾ', 'ਮ']
+  /// ```
+  Iterable<String> splitGurmukhi({
+    bool splitVirama = false,
+    bool extensions = false,
+  }) sync* {
+    var br = 0; // starting index of gurmukhi letter break
+
+    for (var i = 0; i < length; i++) {
+      final curr = codeUnitAt(i);
+      final next = i + 1 < length ? codeUnitAt(i + 1) : null;
+
+      if ((next != null && _isGurmukhiDiacritic(next)) ||
+          (!splitVirama && _isGurmukhiVirama(curr)) ||
+          (extensions && _isSantLipiYayyaModifier(curr))) {
+        continue;
+      }
+
+      yield substring(br, i + 1);
+      br = i + 1;
+    }
+  }
+
+  /// Unicode normalization of Gurmukhi text.
+  ///
+  /// Fix Matras of 'ੳ', 'ਅ', and 'ੲ':
+  /// ```dart
+  /// 'ਅਾਦਿ'.normalizeGurmukhi();
+  /// // ਆਦਿ
+  /// ```
+  ///
+  /// Fix Order of Diacritics:
+  /// ```dart
+  /// 'ਕੰੁਚਰ'.normalizeGurmukhi();
+  /// // ਕੁੰਚਰ
+  /// ```
+  String normalizeGurmukhi() {
+    return _sanitizeGurmukhi(_sortGurmukhiDiacritics(this));
+  }
+}
+
+bool _isGurmukhiChar(int char) => char >= 0x0A00 && char <= 0x0A7F;
+
+bool _isGurmukhiVirama(int char) => char == 0x0A4D;
+
+bool _isGurmukhiDiacritic(int char) => _getGurmukhiDiacriticOrder(char) != 0;
+
+bool _isSantLipiYayyaModifier(int char) {
+  return (char == 0xA833)  // ꠳
+      || (char == 0xA834)  // ꠴
+      || (char == 0xA835); // ꠵
+}
+
+int _getGurmukhiDiacriticOrder(int char) {
+  switch (char) {
+    /// nukta
+    case 0x0A3C: // ਼
+      return -2;
+
+    /// virama/halant
+    case 0x0A4D: // ੍
+      return -1;
+
+    /// signs bottom
+    case 0x0A51: // ੑ
+    case 0x0A75: // ੵ
+      return 1;
```

Comment:
Any reason for separate groups for these?

Reply:
Virama should be before the secondary consonant, and the signs should be after them

Reply:
Going to resolve this and talk about the switch vs index method below